arch: riscv: stacktrace: support stacktrace in early system init

ycsin opened the pull request:
Add support for stacktrace in dummy thread which is used to run
the early system initialization code before the kernel switches
to the main thread.

On RISC-V, the dummy thread will be running temporarily on the
interrupt stack, but currently we do not initialize the stack
info for the dummy thread, hence check the address against the
interrupt stack.

___

before:
```
ASSERTION FAIL [!k_is_pre_kernel()] @ ZEPHYR_BASE/include/zephyr/kernel.h:736
       k_current_get called pre-kernel
[0101_000000.000] <err> os: 
[0101_000000.000] <err> os:  mcause: 11, Environment call from M-mode
[0101_000000.000] <err> os:   mtval: 0
[0101_000000.000] <err> os:      a0: 0000000000000000    t0: 0000000000000000
[0101_000000.000] <err> os:      a1: 0000000000000000    t1: 0000000000000000
[0101_000000.000] <err> os:      a2: 0000000000000000    t2: 0000000000000000
[0101_000000.000] <err> os:      a3: 0000000000000000    t3: 0000000000000000
[0101_000000.000] <err> os:      a4: 0000000000000000    t4: 0000000000000000
[0101_000000.000] <err> os:      a5: 0000000000000000    t5: 0000000000000000
[0101_000000.000] <err> os:      a6: 0000000000000000    t6: 0000000000000000
[0101_000000.000] <err> os:      a7: 0000000000000000
[0101_000000.000] <err> os:      sp: 0000000000000000
[0101_000000.000] <err> os:      ra: 0000000000000000
[0101_000000.000] <err> os:    mepc: 0000000000000000
[0101_000000.000] <err> os: mstatus: 0000000000000000
[0101_000000.000] <err> os: 
[0101_000000.000] <err> os:      s0: 0000000000000000    s6: 0000000000000000
[0101_000000.000] <err> os:      s1: 0000000000000000    s7: 0000000000000000
[0101_000000.000] <err> os:      s2: 0000000000000000    s8: 0000000000000000
[0101_000000.000] <err> os:      s3: 0000000000000000    s9: 0000000000000000
[0101_000000.000] <err> os:      s4: 0000000000000000   s10: 0000000000000000
[0101_000000.000] <err> os:      s5: 0000000000000000   s11: 0000000000000000
[0101_000000.000] <err> os: 
[0101_000000.000] <err> os: call trace:
[0101_000000.000] <err> os: 
[0101_000000.000] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
[0101_000000.000] <err> os: Current thread: 0x00000000 (unknown)
[0101_000000.000] <err> os:  mcause: 11, Environment call from M-mode
[0101_000000.000] <err> os: mdcause: 0, unknown
[0101_000000.000] <err> os:   mtval: 0
[0101_000000.000] <err> os: 
[0101_000000.000] <err> os: Halting system
```

after:
```
ASSERTION FAIL [!k_is_pre_kernel()] @ ZEPHYR_BASE/include/zephyr/kernel.h:736
       k_current_get called pre-kernel
[0101_000000.000] <err> os: 
[0101_000000.000] <err> os:  mcause: 11, Environment call from M-mode
[0101_000000.000] <err> os:   mtval: 0
[0101_000000.000] <err> os:      a0: 0000000000000000    t0: 0000000000000000
[0101_000000.000] <err> os:      a1: 0000000000000000    t1: 0000000000000000
[0101_000000.000] <err> os:      a2: 0000000000000000    t2: 0000000000000000
[0101_000000.000] <err> os:      a3: 0000000000000000    t3: 0000000000000000
[0101_000000.000] <err> os:      a4: 0000000000000000    t4: 0000000000000000
[0101_000000.000] <err> os:      a5: 0000000000000000    t5: 0000000000000000
[0101_000000.000] <err> os:      a6: 0000000000000000    t6: 0000000000000000
[0101_000000.000] <err> os:      a7: 0000000000000000
[0101_000000.000] <err> os:      sp: 0000000000000000
[0101_000000.000] <err> os:      ra: 0000000000000000
[0101_000000.000] <err> os:    mepc: 0000000000000000
[0101_000000.000] <err> os: mstatus: 0000000000000000
[0101_000000.000] <err> os: 
[0101_000000.000] <err> os:      s0: 0000000000000000    s6: 0000000000000000
[0101_000000.000] <err> os:      s1: 0000000000000000    s7: 0000000000000000
[0101_000000.000] <err> os:      s2: 0000000000000000    s8: 0000000000000000
[0101_000000.000] <err> os:      s3: 0000000000000000    s9: 0000000000000000
[0101_000000.000] <err> os:      s4: 0000000000000000   s10: 0000000000000000
[0101_000000.000] <err> os:      s5: 0000000000000000   s11: 0000000000000000
[0101_000000.000] <err> os: 
[0101_000000.000] <err> os: call trace:
[0101_000000.000] <err> os:       0: fp: 0000000000000000 ra: 0000000000000000 [assert_post_action+0x10]
[0101_000000.000] <err> os:       1: fp: 0000000000000000 ra: 0000000000000000 [k_current_get+0x4e]
[0101_000000.000] <err> os:       2: fp: 0000000000000000 ra: 0000000000000000 [k_current_get+0x4e]
[0101_000000.000] <err> os:       3: fp: 0000000000000000 ra: 0000000000000000 [sys_trace_thread_switched_out_user+0xc]
[0101_000000.000] <err> os:       4: fp: 0000000000000000 ra: 0000000000000000 [z_cstart+0x1e0]
[0101_000000.000] <err> os: 
[0101_000000.000] <err> os: >>> ZEPHYR FATAL ERROR 4: Kernel panic on CPU 0
[0101_000000.000] <err> os: Current thread: 0x00000000 (unknown)
[0101_000000.000] <err> os:  mcause: 11, Environment call from M-mode
[0101_000000.000] <err> os: mdcause: 0, unknown
[0101_000000.000] <err> os:   mtval: 0
[0101_000000.000] <err> os: 
[0101_000000.000] <err> os: Halting system
```

ycsin committed:
arch: riscv: stacktrace: support stacktrace in early system init

Add support for stacktrace in dummy thread which is used to run
the early system initialization code before the kernel switches
to the main thread.

On RISC-V, the dummy thread will be running temporarily on the
interrupt stack, but currently we do not initialize the stack
info for the dummy thread, hence check the address against the
interrupt stack.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

npitre reviewed (commented):
You might want to use `is_thread_dummy(thread)` instead.

Yet... this is legitimate only at boot time. Otherwise, after boot,
the dummy thread is masquerading the real thread that is dying andusing
IRQ stack isn't right in that case.

Why not simply initializing `_thread_dummy.stack_info.start` and
`_thread_dummy.stack_info.size` after the call to `z_dummy_thread_init()`
in kernel/init.c ?

ycsin commented:
> You might want to use `is_thread_dummy(thread)` instead.

Didn't know that function exists, removed the compiler guards so that it is always available.

> Yet... this is legitimate only at boot time. Otherwise, after boot,
the dummy thread is masquerading the real thread that is dying and using
IRQ stack isn't right in that case.

Yeah, this PR is basically special handling the dummy thread during init, my understanding is that once we jump to `main` thread, the dummy thread is dead and basically not used anymore so this patch should be fine?

> Why not simply initializing `_thread_dummy.stack_info.start` and `_thread_dummy.stack_info.size` after the call to `z_dummy_thread_init()` in kernel/init.c ?

Each architecture has different names for the stack they used for the dummy thread, I find handling that with `#ifdef`s in `z_dummy_thread_init()` kinda messy unless I create new inline functions that every arch has to implement (i.e. `z_dummy_thread_stack_init(uintptr_t start, size_t size)`, so in the end I decided to take this shortcut instead

ycsin committed:
kernel: thread: constify thread arg of read-only functions

Since these helper functions are read-only, mark the `thread`
arg as `const` so that we can pass const thread to it without
triggering warnings.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>
Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>
